expr/shaper.go

```diff
@@ -336,19 +336,32 @@ func (c *Shaper) fillRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
 			names = append(names, specCol.Name)
 		}
 	}
+	// Now, figure out filled type for the fields.
 	for _, name := range names {
-		ind, _ := spec.ColumnOfField(name)
-		specCol := spec.Columns[ind]
-		if !input.HasField(name) {
+		var inCol, specCol zng.Column
+		i, inputOk := input.ColumnOfField(name)
+		if inputOk {
+			inCol = input.Columns[i]
+		}
+		i, specOk := spec.ColumnOfField(name)
+		if specOk {
+			specCol = spec.Columns[i]
+		}
+		if !inputOk {
 			// Field not in input: fill.
 			cols = append(cols, specCol)
 			continue
 		}
-		ind, _ = input.ColumnOfField(name)
-		inCol := input.Columns[ind]
+		if !specOk {
+			// Field not in spec: input passes through.
+			cols = append(cols, inCol)
+			continue
+		}
+
+		// Field is present both in input and spec: recurse if
+		// both records, or select appropriate type if not.
```

Comment:
I do think this would be a bit simpler without `names`.
```go
	cols := make([]zng.Column, len(input.Columns), len(input.Columns)+len(spec.Columns))
	copy(cols, input.Columns)
	for _, specCol := range spec.Columns {
		if i, ok := input.ColumnOfField(specCol.name); ok {
			// Field is present both in input and spec: recurse if
			// both records, or select appropriate type if not.
			if specRecType, ok := specCol.Type.(*zng.TypeRecord); ok {
				if inRecType, ok := input.Columns[i].Type.(*zng.TypeRecord); ok {
					var err error
					cols.Columns[i], err = c.fillRecordType(inRecType, specRecType)
					if err != nil {
						return nil, err
					}
				} else {
					cols.Columns[i] = specCol
				}
			}
		} else {
			cols = append(cols, specCol)
		}
	}
```

Reply:
oh yes! 😍